### ISSUE TITLE
Custom hooks before and after Rancher draining a node when upgrading RKE2

### ISSUE THREAD

**Issue by bk201** [rancher/rancher#35464]
**Is your feature request related to a problem? Please describe.**
We are currently working on the [Harvester upgrade feature ](https://hackmd.io/@bk201/H1rTAGaLt) and the overall flow looks like this:

![upgrade-Overview drawio](https://user-images.githubusercontent.com/1691518/141310151-051e8bcc-1281-497b-86e1-46c7b8b6fbb2.png)

In phase 2, we upgrade nodes one by one. To make VMs have no downtime, we live-migrate VMs and drain a node before rebooting a node.

The RKE2 cluster in cluster is managed by the embedded Rancher, so in phase 4 we ask Rancher to upgrade the RKE2 runtime by editing cluster resource:

```bash
kubectl edit clusters.provisioning.cattle.io local -n fleet-local
# update spec.kubernetesVersion to newer version
# update spec.rkeConfig for drain options
```

Here we can ask Rancher to drain or not to drain the node before upgrading RKE2 binary on a node:

- Do not drain
    - Restarting`rke2-server` or `rke2-agent` services might cause nodes entering `NotReady` status. Due to a [known issue](https://github.com/longhorn/longhorn/issues/2650) of Longhorn, volumes on nodes will fail and cause all VMs to terminate.  So this is not a valid option for us.
- Drain
    - We can ask Rancher to drain nodes again before upgrading RKE2. This introduces another live migration and draining operation.

**Describe the solution you'd like**

Allow adding custom hooks before and after Rancher draining a node when upgrading RKE2 on nodes. We can combine Phase 2 and Phase 4 into a single-phase and avoid double live migration and draining operations.

**Describe alternatives you've considered**

Upgrading RKE2 in Phase 2. But I don't think it's a good idea because the RKE2 cluster is managed by Rancher server.

**Additional context**

**Comment by yasker** [comment]
Regarding `Do not drain` option, even without the Longhorn issue, `draining` is required at least for Kubernetes minor version upgrade (https://kubernetes.io/docs/tasks/administer-cluster/cluster-upgrade/#manual-deployments and https://kubernetes.io/docs/tasks/administer-cluster/kubeadm/kubeadm-upgrade/#additional-information). So we cannot assume that we don't need to drain the node before the upgrade.

**Comment by ibuildthecloud** [comment]
What I would propose as a fix is this.  We add the following to the provisioning cluster object
```yaml
kind: Cluster
spec:
  rkeConfig:
    upgradeStrategy:
       controlPlaneDrainOptions:
         preDrainHook:
         - annotation: harvester.cattle.io/pre-hook
         postDrainHook:
         - annotation: harvester.cattle.io/pre-hook
       workerDrainOptions:
         preDrainHook:
         - annotation: harvester.cattle.io/post-hook
         postDrainHook:
         - annotation: harvester.cattle.io/post-hook
```

The harvestor controller would then watch capi Machine objects looking for the annotation `rke.cattle.io/pre-drain`.  The value of this annotation is opaque and effectively a random value.  If you see this annotation set you then do the pre-drain behavior and then when done set `harvester.cattle.io/pre-hook` to the value of `rke.cattle.io/pre-drain`.  `harvester.cattle.io/pre-hook` is the annotation name you configured in the provisioning cluster.  The same logic would apply to post hook but just using the `rke.catle.io/post-drain` annotation.

**Comment by yasker** [comment]
@ibuildthecloud once the drain is done, can we just remove the annotation? I am thinking what if there are multiple controllers who want to use the hook in the future. Though we will need an agreed pattern for the hook annotation then.

**Comment by ibuildthecloud** [comment]
@yasker The intention would be that you have one annotation per controller.  In the API I proposed is a list, so more annotation can be added.

**Comment by yasker** [comment]
@ibuildthecloud I see. so it would like this:
1. When a node is going to be drained, `rke.cattle.io/pre-drain` will be set on the CAPI machine object, with a random value.
    * e.g. `rke.cattle.io/pre-drain=123`
2. Rancher will compare the value of the `rke.cattle.io/pre-drain` to the upgradeStrategy
    1. If it's different, Rancher will continue to wait for the pre-drain operation to be done.
    2. Once the value of `rke.cattle.io/pre-drain=harvester.cattle.io/pre-hook` a.k.a matched the value in the upgradeStrategy, Rancher will reset the annotation value and continue with the next step.
        * might need to pay attention to the order since it's a list
3. Rancher drained the node.
4. Rancher will check the value of `rke.cattle.io/post-drain` and perform the same logic as pre-drain.

Does this sound right?

**Comment by ibuildthecloud** [comment]
No.  Rancher will set `rke.cattle.io/pre-drain=123` and the harvester will do it's work and set `harvester.cattle.io/pre-hook=123` and then Rancher will know that hook is done and can continue.

**Comment by yasker** [comment]
@ibuildthecloud updated:
1. When a node is going to be drained, `rke.cattle.io/pre-drain` will be set on the CAPI machine object, with a random value.
    * e.g. `rke.cattle.io/pre-drain=123`
2. Rancher will check the value of the upgradeStrategy's `preDrainHook.annotation`.
    1. If it's non-empty, Rancher will extract the value, e.g. `harvester.cattle.io/pre-hook`, and will wait for `harvester.cattle.io/pre-hook=123` in the object's annotation before continuing.
3. Once the annotation was found, Rancher will drain the node.
4. Rancher will check the value of `postDrainHook.annotation` and perform the same logic as pre-drain.

When will Rancher clear the annotations like `harvester.cattle.io/pre-hook=123`?

**Comment by ibuildthecloud** [comment]
@yasker After the full drain procedure is done or canceled all related annotation `rke.cattle.io/pre-drain`, `harvester.cattle.io/pre-hook`, etc will be deleted.  It's important that the harvester controller not do it's logic the value of `rke.cattle.io/pre-drain` equals the value of  `harvester.cattle.io/pre-hook`.

**Comment by yasker** [comment]
> It's important that the harvester controller not do it's logic the value of `rke.cattle.io/pre-drain` equals the value of `harvester.cattle.io/pre-hook`.

@ibuildthecloud did you mean `not to do it's logic if the value of`? Yeah that makes sense.

**Comment by bk201** [comment]
Thanks @ibuildthecloud and @yasker . So basically the flow might look like this:

![upgrade-Page-4 drawio](https://user-images.githubusercontent.com/1691518/141441201-07f00a1a-abd4-4e04-a46b-f93ee3e817d0.png)


What if the RKE2 version is not changed? Is it possible to trigger this flow?

**Comment by ibuildthecloud** [comment]
We currently drain whenever we update the node plan for a host.  If you want to force a rolling drain of the whole cluster you can do something like like assign a label to all nodes and change the value.  For example

```yaml
kind: Cluster
apiVersion: provisioning.cattle.io/v1

spec:
  rkeConfig:
    machineGlobalConfig:
      node-label+:
      - harvestor.cattle.io/drain-generation=1
```
Then just update drain-generation to 2 and it should do a rolling update according to the update policy of the cluster.

**Comment by bk201** [comment]
@ibuildthecloud Thanks for the tip!

**Comment by ibuildthecloud** [comment]
@bk201 @yasker I need a quick confirmation.    These hooks are for pre-post draining.  So if for example we are upgrading rke2 version we would do

1. pre-drain
2. cordon
3. drain
3. post-drain
4. upgrade rke2
5. uncordon

Is this what you were expecting.  For example, is there a need to do something after rke2 upgrade?  I right all operations are before rke2 because this is just pre/post-drain hooks which is before the actual node operation.

**Comment by ibuildthecloud** [comment]
I just seems more logic to me that you might want pre-drain and post-upgrade.

**Comment by bk201** [comment]
> @bk201 @yasker I need a quick confirmation. These hooks are for pre-post draining. So if for example we are upgrading rke2 version we would do
> 
> 1. pre-drain
> 2. cordon
> 3. drain
> 4. post-drain
> 5. upgrade rke2
> 6. uncordon
> 
> Is this what you were expecting. For example, is there a need to do something after rke2 upgrade? Right now all operations are before rke2 because this is just pre/post-drain hooks which is before the actual node operation.

@ibuildthecloud I think you raise a good point. `post-upgrade` is much useful because we can know a node is ready (the RKE2 process won't restart again).

Currently, I upgrade/reboot OS in SUC jobs' `upgrade` container. So it's a post-drain operation. But maybe it's better to do this after RKE2 upgrade.

BTW, can I still schedule jobs to a cordon node in either `post-drain` or `post-ugprade` stage? By using an SUC jobs, the job pod is not evicted so the `upgrade` container can run after cordoning.

**Comment by yasker** [comment]
@bk201 Can you help with testing?

**Comment by bk201** [comment]
@ibuildthecloud I tested this with a 4-node cluster (3 servers + 1 agent). It works well on the first node (`node1`):
- See annotation `rke.cattle.io/pre-drain` and set its value to `harvester.cattle.io/pre-hook`.
- The node is drained and RKE2 is upgraded.
- See annotation `rke.cattle.io/post-drain` and set its value to `harvester.cattle.io/post-hook`.
- The node is un-cordoned.

For the second server node (`node4`):
- See annotation `rke.cattle.io/pre-drain` and set its value to `harvester.cattle.io/pre-hook`.

But after that, another server node (`node3`) is also cordoned (its machine has `rke.cattle.io/pre-drain` set).
`node4` will be drained and upgraded.
```
╰─$ kubectl get nodes
NAME    STATUS                     ROLES                              AGE   VERSION
node1   Ready                      control-plane,etcd,master,worker   41m   v1.21.6+rke2r1
node2   Ready                      worker                             36m   v1.21.5+rke2r1
node3   Ready,SchedulingDisabled   control-plane,etcd,master,worker   36m   v1.21.5+rke2r1
node4   Ready,SchedulingDisabled   control-plane,etcd,master,worker   36m   v1.21.6+rke2r1
```


I upgrade RKE2 with the config:
```
spec:
  kubernetesVersion: v1.21.6+rke2r1
  localClusterAuthEndpoint: {}
  rkeConfig:
    upgradeStrategy:
      controlPlaneConcurrency: "1"
      workerConcurrency: "1"
      controlPlaneDrainOptions:
        deleteEmptyDirData: true
        enabled: true
        force: true
        ignoreDaemonSets: true
        preDrainHooks:
          - annotation: harvester.cattle.io/pre-hook
        postDrainHooks:
          - annotation: harvester.cattle.io/post-hook
      workerDrainOptions:
        deleteEmptyDirData: true
        enabled: true
        force: true
        ignoreDaemonSets: true
        preDrainHooks:
          - annotation: harvester.cattle.io/pre-hook
        postDrainHooks:
          - annotation: harvester.cattle.io/post-hook
```

---
I also test on a (1 server + 3 agent) scenario. The first server node works as usual, but after the first node is un-cordoned. All the rest agents are cordoned.
```
$ kubectl get nodes
NAME    STATUS                     ROLES                              AGE   VERSION
node1   Ready                      control-plane,etcd,master,worker   14m   v1.21.6+rke2r1
node2   Ready,SchedulingDisabled   worker                             11m   v1.21.5+rke2r1
node3   Ready,SchedulingDisabled   worker                             11m   v1.21.5+rke2r1
node4   Ready,SchedulingDisabled   worker                             11m   v1.21.5+rke2r1
```

**Comment by bk201** [comment]
I tested with `v2.6-28f6a7b29817f0cd002f268fdbbbba4c70f99a58-linux-amd64` and the feature works well.

**Comment by bk201** [comment]
@ibuildthecloud We still see multiple nodes become `SchedulingDisabled` at the same time occasionally, any things or logs I can collect if running into this again? Thanks!

```
harvester-node-0:~ # kubectl get nodes
NAME               STATUS                     ROLES                       AGE     VERSION
harvester-node-0   Ready                      control-plane,etcd,master   8h      v1.22.4+rke2r2
harvester-node-1   Ready                      control-plane,etcd,master   7h56m   v1.22.4+rke2r2
harvester-node-2   Ready,SchedulingDisabled   control-plane,etcd,master   7h43m   v1.22.4+rke2r2
harvester-node-3   Ready,SchedulingDisabled   <none>                      7h27m   v1.22.4+rke2r2
```

**Comment by yasker** [comment]
@bk201 let's file a new issue with v2.6.4 milestone for the bug.

**Comment by bk201** [comment]
> [BUG] Zero downtime upgrade stuck in "Virt launcher pod is terminating" caused by VM crashed (4 nodes) harvester/harvester#1717

Create a new issue to track this: https://github.com/rancher/rancher/issues/35999
